src/test/java/com/google/sps/servlets/RentPostsServletTest.java

```diff
@@ -0,0 +1,160 @@
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.*;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions.Builder;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.WriteResult;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public class RentPostsServletTest {
+  @Mock CollectionReference collectionMock;
+  @Mock Firestore dbMock;
+  @Mock HttpServletRequest request;
+  @Mock HttpServletResponse response;
+  private RentPostsServlet rentPostsServlet;
+
+  @Before
+  public void setUp() throws Exception {    
+    MockitoAnnotations.initMocks(this);
+    rentPostsServlet = new RentPostsServlet();
+    rentPostsServlet.init();
+    FirestoreUtil.setDatabase(dbMock);
+  }
+
+  @Test
+  public void testDatabaseMock_MockExistsAndIsUsedByServlet() throws Exception {
+    assertNotNull(FirestoreUtil.getDatabase());
+    assertEquals(dbMock, FirestoreUtil.getDatabase());
+  }
+
+  @Test
+  public void testPost_postsSingleRentPost() throws Exception {
+    String description = "Test description";
+    String endDate = "2020-07-10";
+    String leaseType = "YEAR_LONG";
+    String numRooms = "1";
+    String price = "100";
+    String roomType = "SINGLE";
+    String startDate = "2020-07-10";
+    String title = "Test title";
+    setupDatabaseMocks();
+    setRequestParameters(description, endDate, leaseType, numRooms,
+      price, roomType, startDate, title);
+    Map<String, Object> expectedData = getExpectedRentData(description, endDate,
+      leaseType, numRooms, price, roomType, startDate, title);
+
+    rentPostsServlet.doPost(request, response);
+
+    verify(dbMock, Mockito.times(1)).collection(RENT_COLLECTION_NAME);
+    verify(collectionMock, Mockito.times(1)).add(expectedData);
+  }
+
+  @Test
+  public void testPost_requestHasUnparsableDates_postsWithNullDates() throws Exception {
```

Comment:
The behavior itself seems questionable, as discussed earlier. 

This point is somewhat debatable: you could say testing it here is extraneous, since it should be covered by tests of RentPost. On the other hand, this is a strong API boundary so it's good test to have...it's just closer to a mini-integration test than a true unit test.

Reply:
Removed ability to post with null dates for the reason discussed above (don't want a listing to be posted to the database without a valid date) and as a result removed the test from the servlets test section.